rtshell/rtwatch.py

```diff
@@ -0,0 +1,228 @@
+    'RTC_STATUS': ('rtc_status', rtc_status_cb),
+    'COMPONENT_PROFILE': ('component_profile', component_profile_cb),
+    'EC_EVENT': ('ec_event', ec_event_cb),
+    'PORT_EVENT': ('port_event', port_event_cb),
+    'CONFIG_EVENT': ('config_event', config_event_cb),
+    'FSM_EVENT': ('fsm_event', fsm_event_cb),
+    'HEARTBEAT': ('heartbeat', heartbeat_cb),
+}
+
+def clean_events(rtcs):
+    # Remove all the event hooks that were added
+    for i in rtcs:
+        i.dynamic = False
+
+def print_logs(paths, options, tree=None):
+    global counter
+    for p in paths:
+        path, port = rtctree.path.parse_path(p[1])
+        if port:
+            raise rts_exceptions.NotAComponentError(p[0])
+        if not path[-1]:
+            raise rts_exceptions.NotAComponentError(p[0])
+        p.append(path)
+
+    if not tree:
+        parsed = [p[2] for p in paths]
+        tree = rtctree.tree.RTCTree(paths=parsed, filter=parsed)
+
+    rtcs = []
+    for p in paths:
+        if not tree.has_path(p[2]):
+            raise rts_exceptions.NoSuchObjectError(p[0])
+        rtc = tree.get_node(p[2])
+        if rtc.is_zombie:
+            raise rts_exceptions.ZombieObjectError(p[0])
+        if not rtc.is_component:
+            raise rts_exceptions.NotAComponentError(p[0])
+
+        rtc.dynamic = True
+        if len(options.filters) == 0 or 'ALL' in options.filters:
+            for (k,v) in filtermap.items():
+                rtc.add_callback(v[0], v[1])
+        else:
+            for f in options.filters:
+                try:
+                    v = filtermap[f]
+                    rtc.add_callback(v[0], v[1])
+                except KeyError:
+                    print('Unknown filter: {0}'.format(f))
+        rtcs.append(rtc)
+
+    # Wait for a keyboard interrupt
+    counter = 0
+    atexit.register(clean_events, rtcs)
+    while True:
+        if options.number > 0 and counter >= options.number:
+            break
+        time.sleep(0.5)
+    clean_events(rtcs)
+
+def main(argv=None, tree=None):
+    usage = '''Usage: %prog [options] <path 1> [path 2...]
+Watch a component event.'''
+    version = rtshell.RTSH_VERSION
+    parser = optparse.OptionParser(usage=usage, version=version)
+    parser.add_option('-n', '--number', dest='number', action='store',
+            type='int', default=-1,
+            help='Number of event to capture. [Default: %default]')
+    parser.add_option('-f', '--filter', dest='filters', action='append',
+            type='string', default=[],
+            help='Event source filters. [Default: ALL]')
+    parser.add_option('-v', '--verbose', dest='verbose', action='store_true',
+            default=False,
+            help='Output verbose information. [Default: %default]')
+
+    if argv:
+        sys.argv = [sys.argv[0]] + argv
+    try:
+        options, args = parser.parse_args()
+    except optparse.OptionError as e:
+        print('OptionError:', e, file=sys.stderr)
+        return 1
+
+    if not args:
```

Comment:
The argparse module supports positional arguments (see https://docs.python.org/2/library/argparse.html?highlight=r#name-or-flags). These should be used instead of manually processing the left-over args.